### DISCUSSION TITLE
Refresh screen fields on option picker change from relationship

### DISCUSSION THREAD

**Discussion by gerhardmpl** [Budibase/budibase#4699] [Help] [answered]
I have two tables `customers` and `discounts` linked by a relation on the column `discount level` of `discounts`. One customer has one current discount level.

When creating or updating a customer row, I want to have an option picker to choose/change the discount level and display the minimum sales required, percentage and term of payment associated with the discount level (defined in the `discounts` table). Using 

```
Customer Data Provider
    Customer Repeater
        Form
           … some customer fields
           Discounts Data Provider
              Discount Picker
              Discount Repeater
                   minimum sale: text field from discount (disabled)
                   percentage: number field from discount (disabled)
                   term of payment: text field from discount (disabled)
```

I can choose/change the different `discount level` but the sales required, percentage and term of payment fields do not refresh (e.g. when trying out different discount levels while working on a customer row). After saving and returning to the customer row, those field are correct in terms of the related discount level. Any idea how to refresh on option picker change?

**Comment by aelias36** [comment]
Is the `Discounts Data Provider` currently referencing `Customer Repeater`? If so, the values only change after updating the row.

Move the discount picker outside of the `Discounts Data Provider`. Then, make the provider do a lookup using the `form.discount level` field instead of the current row.

**Comment by gerhardmpl** [comment]
@aelias36 Thanks, that essentialy works and minimum sales required, percentage and term of payment fields do refresh when switching `discount level`. 

Unfortunatly the fields do not show on the first page load and instead `No rows found` is shown. Could it be that `Form.Fields.discount_level` is not defined when the page first loads?

**Reply by aelias36** [reply]
Yes, if `Form.Fields.discount_level` is empty then that would happen. (If you're updating an existing row, the field should populate with the current row value.)

You can use a condition to change what happens if the field is empty, or to not show anything in the text/number fields. Or you can modify the repeater to say something like `Select a discount level to see discount details`.

**Reply by gerhardmpl** [reply]
Sorry, I should have mentioned that saved discount level is actually showing on the first page load and I assumed that `Form.Fields.discount_level` is not empty:

![grafik](https://user-images.githubusercontent.com/40722408/155805987-f156735c-0c78-4874-84de-f26fa58ba735.png)

**Reply by aelias36** [reply]
Ah, in that case something is up. I've done something very similar before so this should work.

Could you print out the value of `Form.Fields.discount_level` to a paragraph for debugging? Maybe there's something that changes after you select a new discount level.

**Reply by gerhardmpl** [reply]
Indeed, there is a difference:

First page load: `[{"_id":"ro_ta_f3fc482d6f18470fb8e20775b0a6c983_76439b4910cb4c0794c358e0371ee273","primaryDisplay":"50/3/30D"}]`
Change discount: `ro_ta_f3fc482d6f18470fb8e20775b0a6c983_76439b4910cb4c0794c358e0371ee273`

My filter on the Discount Provider is `_id` - `equals` - `binding` - `{{ Form.Fields.discount_level }}` for table `discounts` and that will not work with the value of first page load.

**Reply by aelias36** [reply]
That's strange! I wonder if it's a bug.

I guess the workaround would be to look for one or the other. In javascript that would be something like

```
return $("Form.Fields.discount_level") || $("Form.Fields.discount_level.0._id")
```

**Comment by gerhardmpl** [comment, marked as answer]
Using `return $("Form.Fields.discount_level.0._id");` in the Discount Provider `_id` - `equals` - `binding` - `Javascript function` filter seems to do the trick in both cases. Thanks a lot, that keeps me exploring budibase (despite the somewhat steep learning curve).

**Reply by chrismb23** [reply]
Hi @gerhardmpl - can I ask you if this solution also works for you on first page load for an existing row entry? I'm experiencing that on first page load of an existing row entry I have to manually re-select the value from the relationship picker in roder for it to update the id (checked this by printing the id into a paragraph field).

![image](https://user-images.githubusercontent.com/25127276/200297449-452b5f3c-a23c-4f60-9803-8ab29386d7e0.png)

First page load of existing entry:

![image](https://user-images.githubusercontent.com/25127276/200297340-ed62c99d-e5e3-45b4-abfd-cad401ace6ed.png)

After re-selecting the same value from the relationship picker:

![image](https://user-images.githubusercontent.com/25127276/200297291-79028139-20db-4528-8988-94098917446b.png)
